### DISCUSSION TITLE
Term traversal and term parsing

### DISCUSSION THREAD

**Discussion by baierd** [bitwuzla/bitwuzla#16] [Ideas] [answered]
Hello,
i work at the [JavaSMT ](https://github.com/sosy-lab/java-smt) project and we are very interested in adding your solver to the list of supported solvers.
There are some features that we would find very helpful (see below), and i was wondering if there are any plans to add those in the future.

- Term traversal - the ability to traverse into an existing term and inspect it and use substitution on it. 
- Term parsing - the ability to parse only a single (in our case SMTLIB2) term instead of an entire SMTLIB2 file and return the term as a Bitwuzla term of the current Bitwuzla instance.

Thank you for your time and help.
Best,
Daniel

**Comment by mpreiner** [comment]
Hi Daniel,

that's great news!

Regarding term traversal and substitution, Bitwuzla provides [bitwuzla_term_get_children](https://github.com/bitwuzla/bitwuzla/blob/main/src/api/c/bitwuzla.h#L3598) and [bitwuzla_substitute_terms](https://github.com/bitwuzla/bitwuzla/blob/main/src/api/c/bitwuzla.h#L3384). 

Term parsing is currently not supported, but we might be able to export this functionality. How do you usually use this feature?

**Comment by baierd** [comment]
Thank you for clarifying the API. I didn't find `bitwuzla_term_get_children()` in the [API documentation page](https://bitwuzla.github.io/docs/c/api.html) thats why i missed it.

We use term parsing by giving a String to the parsing function, returning the formula parsed. The parsing function declares functions,type declarations, definitions, and assertions , but does not assert the formula automatically. Usually we expect that the dump of a formula is again parseable such that the same formula is returned from the parse function (in the same context as dump usually only returns the assertion part).

Example (pseudocode): 
String toParse = "(declare-fun arr () (Array BitVec BitVec)) +
(declare-fun arr2 () (Array BitVec BitVec)) +
(declare-fun foo (BitVec) BitVec) +
(declare-fun bv () (_ BitVec 4)) +
(declare-fun bv2 () (_ Bitvec 4)) +
(assert (and (= bv bv2) (=arr arr2) (= (foo bv) bv2)))";
BitwuzlaTerm v = bitwuzla_parse_term_smtlib2(bzla, toParse);

Such that if you now assert v and check SAT, it behaves the same as the current parsing. Just with the difference that now you can parse formulas, manipulate them, assert after you have made some changes... and check SAT in the end.

**Comment by aniemetz** [comment]
@baierd you can find the documentation for `bitwuzla_term_get_children()` here: https://bitwuzla.github.io/docs/c/interface.html#c.bitwuzla_term_get_children

**Comment by aniemetz** [comment]
We'll look into adding support for term parsing in the near future. Not sure yet what this would entail, so this is not yet a promise to support it asap, but a promise to look into if it is possible to support it. If it's not immediately possible, it will take a little longer to add support.

**Comment by baierd** [comment, marked as answer]
It is actually possible to parse terms, as long as you modify the SMTLIB2 input a bit (delete (check-sat) etc.) and conjugate all assertions returned via `bitwuzla_get_assertions(bitwuzla_parser_get_bitwuzla());` after a `bitwuzla_parser_parse()`.
